pysmt/substituter.py

```diff
@@ -20,7 +20,73 @@
 import pysmt.walkers
 from pysmt.walkers.generic import handles
 import pysmt.operators as op
-from pysmt.exceptions import PysmtTypeError
+from pysmt.exceptions import PysmtTypeError, PysmtValueError
+
+
+
+class FunctionInterpretation:
+    """This class represents the interpretation of an uninterpreted
+    function symbol and is intended to be used in substitutions.
+
+    For example, let `phi` be the formula `phi = Equals(Function(f,
+    [Int(2), Int(3)]), a)` where `a` is a Symbol of type INT and `f`
+    is an uninterpreted function with two INT parameters that returns
+    INT. A possible interpretation for `f` could be `f(x, y) = x + y`
+    and a model for `a` could be `5`.
+
+    To represent a model in pysmt we simply use a dict from symbols to
+    constant values, but for interpretations we need a map from
+    symbols to instances of FunctionInterpretation. Each instance of
+    FunctionInterpretation represents the interpretation of a
+    function.
+
+    A FunctionInterpretation is represented as a list of symbols, each
+    having the type corresponding to the parameter typres of the
+    function to be interpreted, plus a function body that is an
+    expression that can only depend on the formal parameters. So to
+    represent the interpretation `f(x, y) = x + y` we construct the
+    instance as follows:
+
+    ```
+    x, y = Symbol('x', INT), Symbol('y', INT)
+    FunctionInterpretation([x, y], Plus(x, y))
+    ```
+    """
+
+    def __init__(self, formal_params, function_body, allow_free_vars=False):
+        """Constructor, taking in input the list of formal parameters and the
+        function body.
+
+        The parameter `allow_free_vars` is used to skip the check that
+        the function body has no free variables other than formal
+        parameter and is used in the SmtLib model-validation utility
+        because functions with uninterpreted return value return
+        special symbols (e.g. @val1) in SmtLib.
```

Comment:
We are not enforcing that these symbols must start with `@`, are we?

Reply:
We are not. In the model-validation pipeline this is enforced by the fact that non-@-symbols are not simplified and thus the model is recognized as invalid.

Do you think we shall enforce this here?

Reply:
I think this is ok as is.